src/main/java/com/faforever/client/replay/OnlineReplayVaultController.java

```diff
@@ -226,36 +244,66 @@ private void enterResultState() {
     loadingPane.setVisible(false);
     backButton.setVisible(false);
-    moreButton.setVisible(false);
+    pagination.setVisible(false);
+    firstPageButton.setVisible(false);
+    lastPageButton.setVisible(false);
   }
 
   private void onSearch(SearchConfig searchConfig) {
+    replaySearchType = ReplaySearchType.SEARCH;
+    onFirstPageOpened(searchConfig);
+  }
+
+  private void onPageChange(SearchConfig searchConfig, int page, boolean firstLoad) {
     enterSearchingState();
-    displayReplaysFromSupplier(() -> replayService.findByQuery(searchConfig.getSearchQuery(), MAX_SEARCH_RESULTS, currentPage++, searchConfig.getSortConfig()));
+    switch (replaySearchType) {
+      case SEARCH:
+        displayReplaysFromSupplier(() -> replayService.findByQueryWithMeta(searchConfig.getSearchQuery(), PAGE_SIZE, page, searchConfig.getSortConfig()), firstLoad);
+        break;
+      case OWN:
+        displayReplaysFromSupplier(() -> replayService.getOwnReplaysWithMeta(PAGE_SIZE, page), firstLoad);
+        break;
+      case NEWEST:
+        displayReplaysFromSupplier(() -> replayService.getNewestReplaysWithMeta(PAGE_SIZE, page), firstLoad);
+        break;
+      case HIGHEST_RATED:
+        displayReplaysFromSupplier(() -> replayService.getHighestRatedReplaysWithMeta(PAGE_SIZE, page), firstLoad);
+        break;
+      case PLAYER:
+        displayReplaysFromSupplier(() -> replayService.getReplaysForPlayerWithMeta(playerId, PAGE_SIZE, page, new SortConfig("startTime", SortOrder.DESC)), firstLoad);
+        break;
+    }
+  }
+
+  private void onFirstPageOpened(SearchConfig searchConfig) {
+    onPageChange(searchConfig, 1, true);
+    if (pagination.getCurrentPageIndex() != 0) {
+      pagination.setCurrentPageIndex(0);
+    }
   }
 
   private void displaySearchResult(List<Replay> replays) {
     displaySearchResult(replays, false);
   }
 
   public void onBackButtonClicked() {
-    if (newestReplaysLoaded) {
-      enterResultState();
-    } else {
-      loadPreselectedReplays();
-    }
+    loadPreselectedReplays();
   }
 
   public void onRefreshButtonClicked() {
-    loadPreselectedReplays();
+    if (pagination.isVisible()) {
+      onPageChange(searchController.getLastSearchConfig(), pagination.currentPageIndexProperty().getValue() + 1, false);
+    } else {
+      loadPreselectedReplays();
+    }
   }
 
   private void loadPreselectedReplays() {
     enterSearchingState();
-    replayService.getNewestReplays(TOP_ELEMENT_COUNT, 1)
-        .thenAccept(replays -> populateReplays(replays, newestPane))
-        .thenCompose(aVoid -> replayService.getHighestRatedReplays(TOP_ELEMENT_COUNT, 1).thenAccept(highestRatedReplays -> populateReplays(highestRatedReplays, highestRatedPane)))
-        .thenCompose(aVoid -> replayService.getOwnReplays(TOP_ELEMENT_COUNT, 1).thenAccept(highestRatedReplays -> populateReplays(highestRatedReplays, ownReplaysPane)))
+    replayService.getNewestReplaysWithMeta(TOP_ELEMENT_COUNT, 1)
```

Comment:
How about u get the total page count in the replay service already u dont need the complete metadata anyway right?

Reply:
We decided against that, as we would still need to carry a tuple as far as we are doing since the pagination gets set in the controller.
This also opens up the possibility to use more of the metadata than just the page amount

Reply:
There is never gonna be any other metadata than page counts